### DISCUSSION TITLE
update rootfs partition and boot partition

### DISCUSSION THREAD

**Discussion by ghost** [rauc/rauc#1330] [General]
We are wanting to ota the rootfs partition and the boot partition that contains the kernel image and device trees.
The partitions we are using are shown below:

/dev/mmcblk2p1 == boot_a partition (that contains the kernel and device trees)
/dev/mmcblk2p2 == rootfs_a 
/dev/mmcblk2p3 == rootfs_b 
/dev/mmcblk2p7 == boot_b partition (that contains the kernel and device trees)

The rauc system.conf is shown below:
``` ini
[system]
compatible=imx8mnevk
bootloader=uboot
max-bundle-download-size=800000000
bundle-formats=verity
data-directory=/data/rauc

[keyring]
path=/etc/rauc/ca.cert.pem

[slot.rootfs.0]
device=/dev/mmcblk2p2
type=ext4
bootname=A

[slot.rootfs.1]
device=/dev/mmcblk2p3
type=ext4
bootname=B

[slot.linux.0]
parent=rootfs.0
device=/dev/mmcblk2p1
type=vfat

[slot.linux.1]
parent=rootfs.1
device=/dev/mmcblk2p7
type=vfat
```

The rauc update-bundle.bb is shown below:
``` bitbake
DESCRIPTION = "RAUC bundle generator"

inherit bundle

RAUC_BUNDLE_FORMAT = "verity"

RAUC_BUNDLE_COMPATIBLE = "${MACHINE}"
RAUC_BUNDLE_VERSION = "3.0.6"
RAUC_BUNDLE_DESCRIPTION = "{\"Critical\": 0,\"ReasonCode\": 0,\"Description\": \"rauc halo update bundle\"}"

# RAUC_BUNDLE_SLOTS = "rootfs linux"
RAUC_BUNDLE_SLOTS = "rootfs"

# RAUC_IMAGE_FSTYPE = "tar.bz2 vfat"
RAUC_IMAGE_FSTYPE = "tar.bz2"

RAUC_SLOT_rootfs = "core-image-halo"
RAUC_SLOT_rootfs[fstype] = "ext4"
RAUC_SLOT_rootfs[adaptive] = "block-hash-index"

RAUC_SLOT_linux = "linux-imx"
RAUC_SLOT_linux[type] = "file"
RAUC_SLOT_linux[file] = "Image"
# RAUC_SLOT_boot[fstype] = "fat32"
# RAUC_SLOT_boot[file] = "Image"
# RAUC_SLOT_boot[adaptive] = "block-hash-index"

BUNDLE_ARGS += ' --mksquashfs-args="-b 64k" '

# optimizations for RAUC adaptive method 'block-hash-index'
# rootfs image size must to be 4K-aligned
IMAGE_ROOTFS_ALIGNMENT = "4096"

# ext4 block and inode size should be set to 4K
EXTRA_IMAGECMD_ext4 = "-i 4096 -b 4096"

# vfat block and inode size should be set to 4K
# EXTRA_IMAGECMD_vfat = "-i 4096 -b 4096"

RAUC_KEY_FILE = "${THISDIR}/files/development-1.key.pem"
RAUC_CERT_FILE = "${THISDIR}/files/development-1.cert.pem"
```

From the above when we perform an update; rauc updates the rootfs and update the corresponding boot slot with the kernel image.

But the boot partition that has been updated can not be mounted and and the kernel reports bad file system

Please can you tell us what we are doing wrong; or what we have missed.
We want to use rauc to update the entire rootfs and the entire corresponding boot partition

**Comment by ejoerns** [comment]
@pritpalm-kuk If you define the slot as `vfat` and install an image, the image is expected to be a `vfat` (*file system*) image.

In your case it is just a kernel image, no file system image.

Normally, file systems are not updated by just putting a single file in it. RAUC supports 'files-based' updates only by creating a tar archive which is then extracted in a newly formatted slot. I also wonder how you define where in the vfat the kernel image should be placed?

Thus if you actually need a vfat there, either let yocto build a vfat image or a tar archive containing the kernel.

**Reply by ghost** [reply]
Thank you for the update; so should we be creating a  tar archive that contains the kernel image and device trees?; If so how do we do this?

Also you motioned the following "let yocto build a vfat image or a tar archive containing the kernel"
How do we in yocto build a vfat image or tar archive?

Also please have a look at our wic file below that defines the partitions in yocto

```
part u-boot --source rawcopy --sourceparams="file=imx-boot" --ondisk mmcblk --no-table --align 32
part /boot --source bootimg-partition --ondisk mmcblk --fstype=vfat --label boot --active --align 8192 --size 64
# part / --source bootimg-partition --ondisk mmcblk --fstype=vfat --label boot_A --align 8192 --size 64
part / --source rootfs --ondisk mmcblk --fstype=ext4 --label rootfs_A --fixed-size=2G --align 8192
part / --source rootfs --ondisk mmcblk --fstype=ext4 --label rootfs_B --fixed-size=2G --align 8192
part /data --source rootfs --rootfs-dir=${IMAGE_ROOTFS}/data --ondisk mmcblk --fstype=ext4 --size 100K --label data --align 8192
part /caam --source rootfs --rootfs-dir=${IMAGE_ROOTFS}/caam --ondisk mmcblk --fstype=ext4 --size 10K --label caam --align 8192
# part / --source bootimg-partition --ondisk mmcblk --fstype=vfat --label boot_B --align 8192 --size 64

bootloader --ptable msdos
```

**Reply by ejoerns** [reply]
The simplest solution would probably be to just put kernel and dtb into the rootfs and let u-boot boot it from there.

The wic file already seems to create a prober `boot` file system. But as far as I know using the intermediate wic artifacts still isn't trivial. Maybe someone has done this before?

**Reply by ghost** [reply]
Thank you for the update
How can I put kernel and dtb into the rootfs partition?
I have seen the following: 
IMAGE_INSTALL_append += "kernel-image"
IMAGE_INSTALL_append += "kernel-devicetree"

But how will the u-boot boot find the kernel how to boot it?

**Reply by ejoerns** [reply]
@pritpalm-kuk I guess this is more a u-boot question than a RAUC question then. The `ext4load` command should allow you to load a kernel from an ext4 rootfs.

**Reply by anthonio9** [reply]
@ejoerns at the moment I'm using a `boot.cmd.in` based on the one from `meta-rauc-community/meta-rauc-raspberrypi/`:

```sh
fdt addr ${fdt_addr} && fdt get value bootargs /chosen bootargs

test -n "${BOOT_ORDER}" || setenv BOOT_ORDER "A B"
test -n "${BOOT_A_LEFT}" || setenv BOOT_A_LEFT 3
test -n "${BOOT_B_LEFT}" || setenv BOOT_B_LEFT 3
test -n "${BOOT_DEV}" || setenv BOOT_DEV "mmc 0:1"

setenv bootpart
setenv raucslot

for BOOT_SLOT in "${BOOT_ORDER}"; do
  if test "x${bootpart}" != "x"; then
    # skip remaining slots
  elif test "x${BOOT_SLOT}" = "xA"; then
    if itest ${BOOT_A_LEFT} -gt 0; then
      setexpr BOOT_A_LEFT ${BOOT_A_LEFT} - 1
      echo "Found valid RAUC slot A"
      setenv bootpart "/dev/mmcblk0p2"
      setenv raucslot "A"
      setenv BOOT_DEV "mmc 0:2"
    fi
  elif test "x${BOOT_SLOT}" = "xB"; then
    if itest ${BOOT_B_LEFT} -gt 0; then
      setexpr BOOT_B_LEFT ${BOOT_B_LEFT} - 1
      echo "Found valid RAUC slot B"
      setenv bootpart "/dev/mmcblk0p3"
      setenv raucslot "B"
      setenv BOOT_DEV "mmc 0:3"
    fi
  fi
done

if test -n "${bootpart}"; then
  setenv bootargs "${bootargs} root=${bootpart} rauc.slot=${raucslot}"
  saveenv
else
  echo "No valid RAUC slot found. Resetting tries to 3"
  setenv BOOT_A_LEFT 3
  setenv BOOT_B_LEFT 3
  saveenv
  reset
fi

fatload mmc 0:1 ${kernel_addr_r} @@KERNEL_IMAGETYPE@@
if test ! -e mmc 0:1 uboot.env; then saveenv; fi;
@@KERNEL_BOOTCMD@@ ${kernel_addr_r} - ${fdt_addr}
```
That boots the system, but `/boot` remains the same. 

The other things added to the yocto config are as @pritpalm-kuk has found:
```sh
IMAGE_INSTALL:append = " kernel-image"
IMAGE_INSTALL:append = " kernel-devicetree"
```
I wonder if more changes to the image should be done. It's not just the device trees and the kernel image, I'd like to have the rpi's `cmdline.txt` and `config.txt` updated as well.

**Reply by ghost** [reply]
Hi 

Another way to would be to change the boot partition to ext4 in the .wic file.
then update the rauc  system.conf to have two boot slots

The above should work; but your boot.cmd file needs to be updated to support two boot slots

**Comment by anthonio9** [comment]
Hi @pritpalm-kuk have you figured out this problem? 
I've run into a similar kind of an issue, where my `/boot` does not get updated at all. I do not want to update the bootloader, just the contents of the `/boot` that is a part of the rootfs (or is it the same thing?).

**Comment by ghost** [comment]
Hi

Thanks for the update.
I did resolve the issue.
Basically I put the kernel and device tree in the rootfs partition.
Then change the uboot to boot the kernel from rootfs

I can share the implementation with you if you wish

Regards

Pritpal

On Wed, 24 Apr 2024 at 08:35, anthonio9 ***@***.***> wrote:

> Hi @pritpalm-kuk <https://github.com/pritpalm-kuk> have you figured out
> this problem?
> I've run into a similar kind of an issue, where my /boot does not get
> updated at all. I do not want to update the bootloader, just the contents
> of the /boot that is a part of the rootfs (or is it the same thing?).
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rauc/rauc/discussions/1330#discussioncomment-9209427>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANVR26H3E22MIFE6A3NSIUDY65OFJAVCNFSM6AAAAABB73GVM6VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TEMBZGQZDO>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

-- 
This e-mail is intended only for the person to whom it is addressed and may 
contain confidential information. If an addressing or transmission error 
has misdirected this e-mail, please notify the author by replying to this 
e-mail. If you are not the intended recipient you must not use, disclose, 
print or rely on the contents of this e-mail and should immediately and 
permanently delete it.


Kenwood Limited is a company registered in England 
with registered number 872044. The registered office of Kenwood Limited is 
1 Kenwood Business Park, New Lane, Havant, Hampshire, PO9 2NH.


Kenwood 
Limited may monitor outgoing and incoming e-mails and other 
telecommunications on its e-mail and telecommunications systems for the 
purpose of ensuring compliance with the law and with Kenwood Limited's 
policies.

**Reply by anthonio9** [reply]
I would really appreciate more details!

**Reply by ghost** [reply]
Hi

What is the contents of your boot partition?

**Reply by anthonio9** [reply]
The image is build for a raspberry pi, so anything that RPI needs. A few device trees and device tree overlays, config.txt and cmdline.txt files, uboot related scripts and the Image file.

**Reply by anthonio9** [reply]
I've found that to explicitly instruct yocto to put the kernel image into rootfs below should be used:

```sh
RRECOMMENDS:${KERNEL_PACKAGE_NAME}-base = "kernel-image"
```

Although it seems that it is already the case.

**Reply by anthonio9** [reply]
If you put everything in the rootfs partition, then what is the boot partition needed for? Can't we just get rid of it?

**Comment by ghost** [comment]
in my local.conf i added the following to put the kernel in the rootfs
IMAGE_INSTALL_append += "kernel-image"
IMAGE_INSTALL_append += "kernel-devicetree"

Then i modified the uboot boot process to boot from ext4 and this worked for us
hence we only do rootfs updates

**Reply by anthonio9** [reply]
Can you share which uboot files you modified? Was it `fw_env.config` or the `boot.scr` file (the latter is also used by RAUC to specify the A/B rootfs partitions).

**Reply by anthonio9** [reply]
one more question, the `_append` syntax is rather old, are you running a dunfel yocto distribution?

**Comment by ghost** [comment]
Hi

our yocto is from gatesgarth
i have attached the patch in this email for the uboot

Regards

Pritpal

On Wed, 24 Apr 2024 at 11:29, anthonio9 ***@***.***> wrote:

> one more question, the _append syntax is rather old, are you running a
> dunfel yocto distribution?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rauc/rauc/discussions/1330#discussioncomment-9211529>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANVR26F6DTDIG7SP7LY7GZDY66CSPAVCNFSM6AAAAABB73GVM6VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TEMJRGUZDS>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

-- 
This e-mail is intended only for the person to whom it is addressed and may 
contain confidential information. If an addressing or transmission error 
has misdirected this e-mail, please notify the author by replying to this 
e-mail. If you are not the intended recipient you must not use, disclose, 
print or rely on the contents of this e-mail and should immediately and 
permanently delete it.


Kenwood Limited is a company registered in England 
with registered number 872044. The registered office of Kenwood Limited is 
1 Kenwood Business Park, New Lane, Havant, Hampshire, PO9 2NH.


Kenwood 
Limited may monitor outgoing and incoming e-mails and other 
telecommunications on its e-mail and telecommunications systems for the 
purpose of ensuring compliance with the law and with Kenwood Limited's 
policies.

**Reply by anthonio9** [reply]
Thanks Pritpal, sadly it seems that github does not support attaching files via mail, I think you would need to do it directly via the github discussion website.

Thanks again!

**Reply by anthonio9** [reply]
Hi @pritpalm-kuk, could you paste the patch here in the comment?

**Comment by ghost** [comment]
Hi

Will do on Monday

Regards

Pritpal

On Fri, 26 Apr 2024 at 11:10, anthonio9 ***@***.***> wrote:

> Hi @pritpalm-kuk <https://github.com/pritpalm-kuk>, could you paste the
> patch here in the comment?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rauc/rauc/discussions/1330#discussioncomment-9235394>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANVR26AZ26P2PQNE5YI2FZDY7IRYHAVCNFSM6AAAAABB73GVM6VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TEMZVGM4TI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

-- 
This e-mail is intended only for the person to whom it is addressed and may 
contain confidential information. If an addressing or transmission error 
has misdirected this e-mail, please notify the author by replying to this 
e-mail. If you are not the intended recipient you must not use, disclose, 
print or rely on the contents of this e-mail and should immediately and 
permanently delete it.


Kenwood Limited is a company registered in England 
with registered number 872044. The registered office of Kenwood Limited is 
1 Kenwood Business Park, New Lane, Havant, Hampshire, PO9 2NH.


Kenwood 
Limited may monitor outgoing and incoming e-mails and other 
telecommunications on its e-mail and telecommunications systems for the 
purpose of ensuring compliance with the law and with Kenwood Limited's 
policies.

**Reply by anthonio9** [reply]
Thanks, I will appreciate that!

**Comment by ghost** [comment]
Hi

I have attached my patched in this email

Regards

Pritpal

On Fri, 26 Apr 2024 at 11:10, anthonio9 ***@***.***> wrote:

> Hi @pritpalm-kuk <https://github.com/pritpalm-kuk>, could you paste the
> patch here in the comment?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rauc/rauc/discussions/1330#discussioncomment-9235394>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANVR26AZ26P2PQNE5YI2FZDY7IRYHAVCNFSM6AAAAABB73GVM6VHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4TEMZVGM4TI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

-- 
This e-mail is intended only for the person to whom it is addressed and may 
contain confidential information. If an addressing or transmission error 
has misdirected this e-mail, please notify the author by replying to this 
e-mail. If you are not the intended recipient you must not use, disclose, 
print or rely on the contents of this e-mail and should immediately and 
permanently delete it.


Kenwood Limited is a company registered in England 
with registered number 872044. The registered office of Kenwood Limited is 
1 Kenwood Business Park, New Lane, Havant, Hampshire, PO9 2NH.


Kenwood 
Limited may monitor outgoing and incoming e-mails and other 
telecommunications on its e-mail and telecommunications systems for the 
purpose of ensuring compliance with the law and with Kenwood Limited's 
policies.

**Reply by anthonio9** [reply]
Hi Pritpal , once again it does not seem that github supports including attachements via email, could you send me the patch on my mail: antoni.s.jankowski@gmail.com?

Thanks again!

**Reply by ghost** [reply]
Hi @anthonio9 , did you manage to get the attachment @pritpalm-kuk mentioned ? Would you mind posting it here ? I am working on a similar system and it would help me a lot. Thank you !

**Reply by anthonio9** [reply]
I did get the attachments, however it turned out not to be what I wanted (and I can't find the mail in my mailbox sadly :( ). The update of the boot partition wasn't the brightest idea, the better approach was to store the kernel image and the dtb file in the rootfs partition and then tell `u-boot` the location of those two files. My approach was based on [`br2rauc`](https://github.com/cdsteinkuehler/br2rauc/) project, which I highly recommend, there a lot of good there to take the inspiration from. Also the [`meta-rauc-community`](https://github.com/rauc/meta-rauc-community) yocto meta layer has since been updated with better examples. I think you should reconsider your need to update the boot partition.

**Reply by ghost** [reply]
Thank you for your answer. Indeed your approach seams to be better, I'll look into it.

**Reply by kroghm96** [reply]
Hi @anthonio9, any way you can share your approach? There are a lot of examples at the rauc community but it's hard to know which is useful and which is not
